src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/MetadataUpdater.cs

```diff
@@ -0,0 +1,63 @@
+using System.Runtime.InteropServices;
+
+namespace System.Reflection.Metadata
+{
+    public static class MetadataUpdater
+    {
+        [DllImport(RuntimeHelpers.QCall)]
+        private static extern unsafe void ApplyUpdate(QCallAssembly assembly, byte* metadataDelta, int metadataDeltaLength, byte* ilDelta, int ilDeltaLength, byte* pdbDelta, int pdbDeltaLength);
+
+        /// <summary>
+        /// Updates the specified assembly using the provided metadata, IL and PDB deltas.
+        /// </summary>
+        /// <remarks>
+        /// Currently executing methods will continue to use the existing IL. New executions of modified methods will
+        /// use the new IL. Different runtimes may have different limitations on what kinds of changes are supported,
+        /// and runtimes make no guarantees as to the state of the assembly and process if the delta includes
+        /// unsupported changes.
+        /// </remarks>
+        /// <param name="assembly">The assembly to update.</param>
+        /// <param name="metadataDelta">The metadata changes to be applied.</param>
+        /// <param name="ilDelta">The IL changes to be applied.</param>
+        /// <param name="pdbDelta">The PDB changes to be applied.</param>
+        /// <exception cref="ArgumentNullException">The assembly argument is null.</exception>
+        /// <exception cref="NotSupportedException">The update could not be applied.</exception>
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta)
+        {
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            RuntimeAssembly? runtimeAssembly = assembly as RuntimeAssembly;
+            if (runtimeAssembly == null)
+            {
+                throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);
+            }
+
+            unsafe
+            {
+                RuntimeAssembly rtAsm = runtimeAssembly;
+                fixed (byte* metadataDeltaPtr = metadataDelta, ilDeltaPtr = ilDelta, pdbDeltaPtr = pdbDelta)
+                {
+                    ApplyUpdate(new QCallAssembly(ref rtAsm), metadataDeltaPtr, metadataDelta.Length, ilDeltaPtr, ilDelta.Length, pdbDeltaPtr, pdbDelta.Length);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the metadata update capabilities.
+        /// </summary>
+        public static string GetCapabilities() => "Baseline AddMethodToExistingType AddStaticFieldToExistingType AddInstanceFieldToExistingType NewTypeDefinition";
+
+        /// <summary>
+        /// Returns true if the apply assembly update is enabled and available.
+        /// </summary>
+        public static bool IsSupported => Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "" || Environment.GetEnvironmentVariable("COMPlus_ForceEnc") == "1";
```

Comment:
It is not unusual for the user code to clear env variables like this so that they are not inherited by child processes; or on the other hand to set these env variable so that the child processes have them.

Should this rather return the setting cached by the runtime?

Reply:
`Debugger.IsAttached` can change while the process is running. I do not think it is desirable behavior - it is hard to write reliable code when the `IsSupported` can be changing underneath you.

Reply:
> Should this rather return the setting cached by the runtime?

At a minimum we should be caching this so that the value doesn't change, e.g.
```suggestion
        public static bool IsSupported { get; } = Debugger.IsAttached || Environment.GetEnvironmentVariable("DOTNET_MODIFIABLE_ASSEMBLIES") != "" || Environment.GetEnvironmentVariable("COMPlus_ForceEnc") == "1";
```

But, yes, it'd be better if we could ask the runtime for its view on whether updates could potentially be applied.  I'd suggested that previously and was told there was some reason we couldn't, but I don't remember what it was.

Reply:
Yes, I forgot about caching this value like we talked about.  Hopefully what Stephen suggests is enough. The runtime really can't do any better via a QCall other than reading these as config values which can't be changed by the app.

Reply:
If this reads the env variables independently on the runtime, this property can be out of sync with what runtime thinks. I think it would be better to just return the value that the runtime cached.

Reply:
This roughly what I think the QCall would look like to get the runtime apply update supported state. It would still be cached in MetadataUpdater.IsSupported. Is this what you are thinking?
```
BOOL QCALLTYPE AssemblyNative::IsApplyUpdateSupported()
{
    QCALL_CONTRACT;

    BEGIN_QCALL;

#ifdef EnC_SUPPORTED
    BOOL result = CORDebuggerAttached() || g_pConfig->ForceEnc() || g_pConfig->DebugAssembliesModifiable();
#else
    BOOL result = false;
#endif

    END_QCALL;

    return result
}
```

Reply:
Yes, something like this. Why do we need `CORDebuggerAttached` in the condition? I do not think attaching a debugger will turn on EnC today unless it was turned on earlier.

Reply:
It is for the debugger launch case which most debuggers enable ENC for user modules.  This is a case where ENC can be enabled but neither env var has been set.  There really isn't any global state, but per module so CORDebuggerAttached is the closest thing for the launch case.

For debugger attach, either ForceEnc or DebugAssembliesModifiable configs would have to be set to enable ENC.